test/battle/crit_chance.c

```diff
@@ -0,0 +1,252 @@
+{
+    u32 species;
+    u32 ability;
+
+    PARAMETRIZE { species = SPECIES_KINGLER; ability = ABILITY_SHELL_ARMOR; }
+    PARAMETRIZE { species = SPECIES_ARMALDO; ability = ABILITY_BATTLE_ARMOR; }
+
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_SUNSTEEL_STRIKE].ignoresTargetAbility == TRUE);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(species) { Ability(ability); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_SUNSTEEL_STRIKE, criticalHit: TRUE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_SUNSTEEL_STRIKE, player);
+        MESSAGE("A critical hit!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Mold Breaker, Teravolt and Turboblaze ignore Battle Armor and Shell Armor")
+{
+    u32 j;
+    static const u32 pokemonPlayer[][2] =
+    {
+        {SPECIES_PINSIR, ABILITY_MOLD_BREAKER},
+        {SPECIES_ZEKROM, ABILITY_TERAVOLT},
+        {SPECIES_KYUREM_WHITE,  ABILITY_TURBOBLAZE},
+    };
+
+    u32 speciesPlayer;
+    u32 abilityPlayer;
+    u32 speciesOpponent;
+    u32 abilityOpponent;
+
+    for (j = 0; j < ARRAY_COUNT(pokemonPlayer); j++)
+    {
+        PARAMETRIZE {
+            speciesPlayer = pokemonPlayer[j][0];
+            abilityPlayer = pokemonPlayer[j][1];
+            speciesOpponent = SPECIES_KINGLER;
+            abilityOpponent = ABILITY_SHELL_ARMOR;
+        }
+
+        PARAMETRIZE {
+            speciesPlayer = pokemonPlayer[j][0];
+            abilityPlayer = pokemonPlayer[j][1];
+            speciesOpponent = SPECIES_ARMALDO;
+            abilityOpponent = ABILITY_BATTLE_ARMOR;
+        }
+    }
+
+    GIVEN {
+        PLAYER(speciesPlayer) { Ability(abilityPlayer); }
+        OPPONENT(speciesOpponent) { Ability(abilityOpponent); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, criticalHit: TRUE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, player);
+        MESSAGE("A critical hit!");
+    }
+}
+
+SINGLE_BATTLE_TEST("User effected by Laser Focus causes moves to result in a critical hit")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_LASER_FOCUS].effect == EFFECT_LASER_FOCUS);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_LASER_FOCUS); }
+        TURN { MOVE(player, MOVE_TACKLE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_LASER_FOCUS, player);
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, player);
+        MESSAGE("A critical hit!");
+    }
+}
+
+SINGLE_BATTLE_TEST("If the target is poisoned the ability Merciless causes a moev to result in a critical hit")
+{
+    GIVEN {
+        PLAYER(SPECIES_MAREANIE) { Ability(ABILITY_MERCILESS); }
+        OPPONENT(SPECIES_WOBBUFFET) { Status1(STATUS1_POISON); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, player);
+        MESSAGE("A critical hit!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Focus Energy increases the user's critical hit ratio by one stage")
```

Comment:
I think Focus Energy increases the ratio by 2 stages?

Reply:
Yeah, wrong description. Corrections applied.